tools/c7n_azure/c7n_azure/resources/cosmos_db.py

```diff
@@ -52,6 +74,193 @@ class resource_type(ArmResourceManager.resource_type):
+                c.update({'c7n:document-endpoint':
+                         parent_resource.get('properties').get('documentEndpoint')})
+                collections.append(c)
+
+        return collections
+
+
+@CosmosDBCollection.filter_registry.register('offer')
+@CosmosDBDatabase.filter_registry.register('offer')
+class CosmosDBOfferFilter(ValueFilter):
+    schema = type_schema('offer', rinherit=ValueFilter.schema)
+    schema_alias = True
+
+    def process(self, resources, event=None):
+        futures = []
+        results = []
+
+        # Group all resources by account because offers are queried per account not per collection
+        account_sorted = sorted(resources, key=CosmosDBOfferFilter.account_key)
+        account_grouped = [list(it) for k, it in groupby(
+            account_sorted,
+            CosmosDBOfferFilter.account_key)]
+
+        # Process database groups in parallel
+        with self.executor_factory(max_workers=3) as w:
+            for resource_set in account_grouped:
+                futures.append(w.submit(self.process_resource_set, resource_set))
+
+            for f in as_completed(futures):
+                if f.exception():
+                    self.log.warning(
+                        "Diagnostic settings filter error: %s" % f.exception())
+                    continue
+                else:
+                    results.extend(f.result())
+
+            return results
+
+    def process_resource_set(self, resources):
+        matched = []
+
+        try:
+            # Skip if offer key is present anywhere because we already
+            # queried and joined offers in a previous filter instance
+            if not resources[0].get('c7n:offer'):
+
+                # Get the data client keys
+                parent_key = resources[0]['c7n:parent-id']
+                key = CosmosDB.get_cosmos_key(
+                    ResourceIdParser.get_resource_group(parent_key),
+                    ResourceIdParser.get_resource_name(parent_key),
+                    self.manager)
+
+                # Build a data client
+                data_client = CosmosClient(
+                    url_connection=resources[0]['c7n:document-endpoint'],
+                    auth={'masterKey': key})
+
+                # Get the offers
+                offers = list(data_client.ReadOffers())
+
+                # Match up offers to collections
+                for resource in resources:
+                    offer = [o for o in offers if o['resource'] == resource['_self']]
+                    resource['c7n:offer'] = offer
+
+            # Pass each resource through the base filter
+            for resource in resources:
+                filtered_resource = super(CosmosDBOfferFilter, self).process(
+                    resource['c7n:offer'],
+                    event=None)
+
+                if filtered_resource:
+                    matched.append(resource)
+
+        except Exception as error:
+            log.warning(error)
```

Comment:
It is undefined behavior.. If an exception occurs, we ignore a chunk of resources..
I think it needs to be a failfast (raise exception to the ThreadHelper) or handle exception on per-resource basis

Reply:
In this case it isn't just a chunk of resources - it fails a CosmosDB account (the ARM resource) which sort of feels OK.  

We don't really do anything per-resource, we only make one set of API calls per `process-resource-set` which is at the Account level, so all the resources are likely to fail together.